### PR TITLE
added Astraled Mantis its the same as airQualityLuminary_V0_02

### PR DESCRIPTION
**Pull request by J2thatsme**

### CODE REVIEW

File: types/astraled/mantis/uplink.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+function bytesToFloat(bytes) {
+  var bits_float = bytes;
+  var sign = (bits_float >>> 31 === 0) ? 1.0 : -1.0;
+  var e = bits_float >>> 23 & 0xff;
+  var m = (e === 0) ? (bits_float & 0x7fffff) << 1 : (bits_float & 0x7fffff) | 0x800000;
+  var f = sign * m * Math.pow(2, e - 150);
+  return f;
+}
+
+function get_byte(buff, nr) {
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js:23]: use camelCase

File: types/astraled/mantis/uplink.spec.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        assert.equal(value.data.temperature, 23.21);
+        assert.equal(value.data.humidity, 32.6);
+        assert.equal(value.data.voc, 317);
+        assert.equal(value.data.co2, 664);
+
+        validate(value.data, schema, { throwError: true });
+      });
+
+      consume(data);
+      done();
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.spec.js:84]: not necessary to call done

**Review by mariusgiger — COMMENTED** [review]
looks good, some comments

File: types/astraled/mantis/meta.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "Mantis",
+  "version": "1.0.0",
+  "manufacturer": "AstraLED",
+  "url": "https://www.astra-led.ch/produkte/buerostehleuchte/",
+  "description": "AstraLED Indoor LoRaWAN® device for measuring the indoor environment and signaling those through different colors",
+  "author": "Akenza AG",
+  "firmwareVersion": "V0.0.2",
```
[the remark this answers is not in the document]
Comment (J2thatsme) [reply types/astraled/mantis/meta.json:8]: V0.0.2 ? Directly from Astraled

File: types/astraled/airQualityLuminary_V0_02/uplink.js [hunk 4]
```diff
@@ -177,7 +177,7 @@ function consume(event) {
         data.longitude = get_float(buff, rd_pos);
         break;
       case 35:
-        data.light_state = get_byte(buff, rd_pos);
+        data.lightState = get_byte(buff, rd_pos);
```
Comment (mariusgiger) [inline types/astraled/airQualityLuminary_V0_02/uplink.js]: is this change voluntary? if yes change all data keys to camelCase

File: types/astraled/mantis/uplink.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  var m = (e === 0) ? (bits_float & 0x7fffff) << 1 : (bits_float & 0x7fffff) | 0x800000;
+  var f = sign * m * Math.pow(2, e - 150);
+  return f;
+}
+
+function getByte(buff, nr) {
+  return Bits.bitsToUnsigned(buff.substr(nr * 8, 8));
+}
+
+function get_word(buff, nr) {
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js]: 🐫

File: types/astraled/mantis/uplink.js [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+  return swap16(Bits.bitsToUnsigned(buff.substr(nr * 8, 16)));
+}
+
+function get_long(buff, nr) {
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js]: 🐫

File: types/astraled/mantis/uplink.js [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  return swap32(Bits.bitsToUnsigned(buff.substr(nr * 8, 32)));
+}
+
+function get_float(buff, nr) {
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js]: 🐫

Comment (mariusgiger) [reply types/astraled/mantis/uplink.js]: please use a linter to make sure these things pop up during writing code, call me if you need help

**Review by mariusgiger — CHANGES_REQUESTED** [review]

**Comment by J2thatsme** [comment]
@mariusgiger Please review again

File: types/astraled/mantis/uplink.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+function getLong(buff, nr) {
+  return swap32(Bits.bitsToUnsigned(buff.substr(nr * 8, 32)));
+}
+
+function getFloat(buff, nr) {
+  return bytesToFloat(getLong(buff, nr));
+}
+
+function consume(event) {
+
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js:40]: unused newline can be removed

File: types/astraled/mantis/uplink.js [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+  var data_size = payload_hex.length / 2;
+
+  var data = {};
+  var lifecycle = {};
+
+  var rd_pos = 0;
+
+  do {
+    var para_size = getByte(buff, rd_pos++) & 0x3F;  // get size
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js:52]: please also use camleCase for variables

**Review by mariusgiger — CHANGES_REQUESTED** [review]

File: types/astraled/mantis/uplink.js [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+  const dataSize = payload_hex.length / 2;
+
+  const data = {};
+  const lifecycle = {};
+
+  let rdPos = 0;
+
+  do {
+    const paraSize = getByte(buff, rdPos++) & 0x3F;  // get size
```
Comment (mariusgiger) [inline types/astraled/mantis/uplink.js:51]: prevent using abbreviations

File: types/astraled/airQualityLuminary_V0_02/uplink.js [hunk 11]
```diff
@@ -49,166 +49,166 @@ function consume(event) {
...
-        lifecycle.iaq_rgbw_bad_blue = get_byte(buff, rd_pos + 2);
+        lifecycle.iaq_rgbw_bad_red = getByte(buff, rd_pos);
+        lifecycle.iaq_rgbw_bad_green = getByte(buff, rd_pos + 1);
+        lifecycle.iaq_rgbw_bad_blue = getByte(buff, rd_pos + 2);
         break;
       case 28:
-        lifecycle.iaq_rgbw_deadly_red = get_byte(buff, rd_pos);
-        lifecycle.iaq_rgbw_deadly_green = get_byte(buff, rd_pos + 1);
-        lifecycle.iaq_rgbw_deadly_blue = get_byte(buff, rd_pos + 2);
+        lifecycle.iaq_rgbw_deadly_red = getByte(buff, rd_pos);
```
Comment (mariusgiger) [inline types/astraled/airQualityLuminary_V0_02/uplink.js:155]: 🐫

File: CHANGELOG.md [hunk 12]
```diff
@@ -24,6 +24,12 @@ All notable changes to this project will be documented in this file.
 
 ## [1.0.2] - 2021-05-10
 
+### Added
+
+- AstraLED - Mantis
+
+## [1.0.2] - 2021-05-10
```
Comment (mariusgiger) [inline CHANGELOG.md:31]: please correct the version number as there seem to be duplicates

File: types/astraled/airQualityLuminary_V0_02/uplink.js [hunk 13]
```diff
@@ -12,218 +12,217 @@ function swap32(val) {
 }
 
 function bytesToFloat(bytes) {
-  var bits_float = bytes;
-  var sign = (bits_float >>> 31 === 0) ? 1.0 : -1.0;
-  var e = bits_float >>> 23 & 0xff;
-  var m = (e === 0) ? (bits_float & 0x7fffff) << 1 : (bits_float & 0x7fffff) | 0x800000;
-  var f = sign * m * Math.pow(2, e - 150);
+  const bitsFloat = bytes;
```
Comment (mariusgiger) [inline types/astraled/airQualityLuminary_V0_02/uplink.js:15]: why is this reassignment necessary?

File: types/astraled/airQualityLuminary_V0_02/uplink.js [hunk 14]
```diff
@@ -12,218 +12,217 @@ function swap32(val) {
...
-        lifecycle.protocol_version = "Version: " + get_byte(buff, rd_pos + 1) + "." + get_byte(buff, rd_pos);
+        lifecycle.protocolVersion = `Version: ${  getByte(buff, rdPos + 1)  }.${  getByte(buff, rdPos)}`;
         break;
       case 2:
-        lifecycle.app_version = "Version: " + get_byte(buff, rd_pos + 2) + "." + get_byte(buff, rd_pos + 1) + "p" + get_byte(buff, rd_pos);
+        lifecycle.appVersion = `Version: ${  getByte(buff, rdPos + 2)  }.${  getByte(buff, rdPos + 1)  }p${  getByte(buff, rdPos)}`;
         break;
       case 3:
-        lifecycle.lora_stack_version = "Version: " + get_byte(buff, rd_pos + 2) + "." + get_byte(buff, rd_pos + 1) + "." + get_byte(buff, rd_pos);
+        lifecycle.loraStackVersion = `Version: ${  getByte(buff, rdPos + 2)  }.${  getByte(buff, rdPos + 1)  }.${  getByte(buff, rdPos)}`;
```
Comment (mariusgiger) [inline types/astraled/airQualityLuminary_V0_02/uplink.js:62]: there seems to be some strange spaces in here ${  getByte(buff, rdPos + 2)  }, is this voluntary?

**Review by mariusgiger — COMMENTED** [review]
looks good, please correct the version number

File: types/astraled/airQualityLuminary_V0_02/uplink.js (hunk 14 above)
Comment (J2thatsme) [reply types/astraled/airQualityLuminary_V0_02/uplink.js:62]: this was autocorrected

File: CHANGELOG.md [hunk 15]
```diff
@@ -4,30 +4,36 @@ All notable changes to this project will be documented in this file.
...
 - Dragino - LSE01
 
-## [1.0.3] - 2021-05-19
+## [1.0.4] - 2021-05-19
 
 ### Added
 
 - Digitalmatter - Oyster
 
+## [1.0.3] - 2021-05-10
```
Comment (mariusgiger) [inline CHANGELOG.md:31]: this should be the latest version instead of introducing an in-between version..

Comment (mariusgiger) [reply CHANGELOG.md:37]: @J2thatsme please revert changing these below

File: CHANGELOG.md [hunk 16]
```diff
@@ -8,6 +8,7 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 
+- AstraLED - Mantis
```
Comment (mariusgiger) [inline CHANGELOG.md:11]: add a new version for this

**Review by mariusgiger — APPROVED** [review]

**Pull request merged**